pkg/gatewayserver/upstream/ns/ns.go

```diff
@@ -78,6 +78,9 @@ func (h *Handler) ConnectGateway(ctx context.Context, ids ttnpb.GatewayIdentifie
 
 // HandleUp implements upstream.Handler.
 func (h *Handler) HandleUp(ctx context.Context, _ ttnpb.GatewayIdentifiers, ids ttnpb.EndDeviceIdentifiers, msg *ttnpb.GatewayUp) error {
```

Comment:
This signature is wrong; `ttnpb.GatewayUp` can contain multiple `UplinkMessage` from different end devices.

There's no reason to pass `GatewayUp` here; actually it gets wrongly reconstructed in `pkg/gatewayserver/gatewayserver.go` L500. Take `*ttnpb.UplinkMessage` here, rename to `HandleUplink` and introduce a `HandleStatus` that doesn't take any end device identifiers

Reply:
Ok but FTR, this was how I created the Original Offloading PR and I was asked to change to using  `ttnpb.GatewayUp`during a review there.

Reply:
OK, progressive insight maybe!